akka-persistence-query/src/main/scala/akka/persistence/query/internal/QuerySerializer.scala

```diff
@@ -0,0 +1,209 @@
+  private final val NoOffsetManifest = "NO"
+
+  // persistenceId and timestamp must not contain this separator char
+  private val timestampOffsetSeparator = ';'
+
+  override def manifest(o: AnyRef): String = o match {
+    case _: EventEnvelope[_] => EventEnvelopeManifest
+    case offset: Offset      => toStorageRepresentation(offset)._2
+    case _ =>
+      throw new IllegalArgumentException(s"Can't serialize object of type ${o.getClass} in [${getClass.getName}]")
+  }
+
+  override def toBinary(o: AnyRef): Array[Byte] = o match {
+    case env: EventEnvelope[_] =>
+      val builder = QueryMessages.EventEnvelope.newBuilder()
+
+      val (offset, offsetManifest) = toStorageRepresentation(env.offset)
+
+      builder
+        .setPersistenceId(env.persistenceId)
+        .setEntityType(env.entityType)
+        .setSlice(env.slice)
+        .setSequenceNr(env.sequenceNr)
+        .setTimestamp(env.timestamp)
+        .setOffset(offset)
+        .setOffsetManifest(offsetManifest)
+
+      env.eventOption.foreach(event => builder.setEvent(payloadSupport.payloadBuilder(event)))
+      env.eventMetadata.foreach(meta => builder.setMetadata(payloadSupport.payloadBuilder(meta)))
+
+      builder.build().toByteArray()
+
+    case offset: Offset =>
+      toStorageRepresentation(offset)._1.getBytes(UTF_8)
+
+    case _ =>
+      throw new IllegalArgumentException(s"Cannot serialize object of type [${o.getClass.getName}]")
+  }
+
+  override def fromBinary(bytes: Array[Byte], manifest: String): AnyRef = manifest match {
+    case EventEnvelopeManifest =>
+      val env = QueryMessages.EventEnvelope.parseFrom(bytes)
+
+      val offset = fromStorageRepresentation(env.getOffset, env.getOffsetManifest)
+
+      val eventOption =
+        if (env.hasEvent) Option(payloadSupport.deserializePayload(env.getEvent))
+        else None
+      val metaOption =
+        if (env.hasMetadata) Option(payloadSupport.deserializePayload(env.getMetadata))
+        else None
+
+      new EventEnvelope(
+        offset,
+        env.getPersistenceId,
+        env.getSequenceNr,
+        eventOption,
+        env.getTimestamp,
+        metaOption,
+        env.getEntityType,
+        env.getSlice)
+
+    case _ =>
+      fromStorageRepresentation(new String(bytes, UTF_8), manifest)
+  }
+
+  /**
+   * Deserialize an offset from a stored string representation and manifest.
+   * The offset is converted from its string representation to its real type.
+   */
+  private def fromStorageRepresentation(offsetStr: String, manifest: String): Offset = {
+    manifest match {
+      case TimestampOffsetManifest     => timestampOffsetFromStorageRepresentation(offsetStr)
+      case SequenceOffsetManifest      => Offset.sequence(offsetStr.toLong)
+      case TimeBasedUUIDOffsetManifest => Offset.timeBasedUUID(UUID.fromString(offsetStr))
+      case NoOffsetManifest            => NoOffset
+      case _ =>
+        val parts = manifest.split(':')
```

Comment:
this is in the same as the String format used in Akka Projections

Reply:
make it a named constant for extra clarity